src/Simulation/Simulators/SparseSimulator/Native/gates.h

```diff
@@ -0,0 +1,284 @@
+                        return "Y";
+                    case OP::MCY:
+                        return "MCY";
+                    case OP::Z:
+                        return "Z";
+                    case OP::MCZ:
+                        return "MCZ";
+                    case OP::M:
+                        return "M";
+                    case OP::Measure:
+                        return "Measure";
+                    case OP::Exp:
+                        return "Exp";
+                    case OP::MCExp:
+                        return "MCExp";
+                    case OP::R1:
+                        return "R1";
+                    case OP::MCR1:
+                        return "MCR1";
+                    case OP::Rx:
+                        return "Rx";
+                    case OP::Rz:
+                        return "Rz";
+                    case OP::Ry:
+                        return "Ry";
+                    case OP::MCRx:
+                        return "MCRx";
+                    case OP::MCRz:
+                        return "MCRz";
+                    case OP::MCRy:
+                        return "MCRy";
+                    case OP::SWAP:
+                        return "SWAP";
+                    case OP::MCSWAP:
+                        return "MCSWAP";
+                    case OP::Assert:
+                        return "Assert";
+                    case OP::PermuteSmall:
+                        return "Perm_S";
+                    case OP::PermuteLarge:
+                        return "Perm_L";
+                    case OP::Proj:
+                        return "Project";
+                    case OP::MCProj:
+                        return "MCProj";
+                    case OP::Allocate:
+                        return "Alloc";
+                    case OP::Release:
+                        return "Release";
+                    default:
+                        return "Not a gate";
+                }
+            }
+
+            // Used in operation queues for phases/permutations
+            // Different constructors correspond to different 
+            // kinds of gates, so some data is not initialized 
+            // for certain types of gates
+            struct operation {
+                OP gate_type;
+                logical_qubit_id target;
+                operation(OP gate_type_arg,
+                    logical_qubit_id target_arg) :
+                    gate_type(gate_type_arg),
+                    target(target_arg) {}
+
+                std::vector<logical_qubit_id> controls;
+                operation(OP gate_type_arg, 
+                    logical_qubit_id target_arg, 
+                    std::vector<logical_qubit_id> controls_arg
+                    ) : gate_type(gate_type_arg),
+                        target(target_arg),
+                        controls(controls_arg){}
+
+                logical_qubit_id shift;
+                logical_qubit_id target_2;
```

Comment:
FYI: A number of data members stay **un**initialized in a number of constructors.
You may find useful the [Non-static data member initializers](https://github.com/AnthonyCalandra/modern-cpp-features#non-static-data-member-initializers).